src/backend/d3d12/CommandBufferD3D12.cpp

```diff
@@ -16,150 +16,195 @@
-                      {
-                          SetVertexBuffersCmd* cmd = commands.NextCommand<SetVertexBuffersCmd>();
-                      }
-                      break;
-
-                  case Command::TransitionBufferUsage:
-                      {
-                          TransitionBufferUsageCmd* cmd = commands.NextCommand<TransitionBufferUsageCmd>();
-                      }
-                      break;
-
-                  case Command::TransitionTextureUsage:
-                      {
-                          TransitionTextureUsageCmd* cmd = commands.NextCommand<TransitionTextureUsageCmd>();
-                      }
-                      break;
-              }
-          }
-      }
+                case Command::SetBindGroup:
+                    {
+                        SetBindGroupCmd* cmd = commands.NextCommand<SetBindGroupCmd>();
+                    }
+                    break;
+
+                case Command::SetIndexBuffer:
+                    {
+                        SetIndexBufferCmd* cmd = commands.NextCommand<SetIndexBufferCmd>();
+
+                        Buffer* buffer = ToBackend(cmd->buffer.Get());
+                        D3D12_INDEX_BUFFER_VIEW bufferView;
+                        bufferView.BufferLocation = buffer->GetVA() + cmd->offset;
+                        bufferView.SizeInBytes = buffer->GetSize() - cmd->offset;
+                        bufferView.Format = DXGIIndexFormat(cmd->format);
+
+                        commandList->IASetIndexBuffer(&bufferView);
+                    }
+                    break;
+
+                case Command::SetVertexBuffers:
+                    {
+                        SetVertexBuffersCmd* cmd = commands.NextCommand<SetVertexBuffersCmd>();
+                        auto buffers = commands.NextData<Ref<BufferBase>>(cmd->count);
+                        auto offsets = commands.NextData<uint32_t>(cmd->count);
+
+                        auto inputState = ToBackend(lastPipeline->GetInputState());
+
+                        std::array<D3D12_VERTEX_BUFFER_VIEW, kMaxVertexInputs> d3d12BufferViews;
+                        for (uint32_t i = 0; i < cmd->count; ++i) {
+                            auto input = inputState->GetInput(cmd->startSlot + i);
+                            Buffer* buffer = ToBackend(buffers[i].Get());
+                            d3d12BufferViews[i].BufferLocation = buffer->GetVA() + offsets[i];
+                            d3d12BufferViews[i].StrideInBytes = input.stride;
+                            d3d12BufferViews[i].SizeInBytes = buffer->GetSize() - offsets[i];
+                        }
+
+                        commandList->IASetVertexBuffers(cmd->startSlot, cmd->count, d3d12BufferViews.data());
+                    }
+                    break;
+
+                case Command::TransitionBufferUsage:
+                    {
+                        TransitionBufferUsageCmd* cmd = commands.NextCommand<TransitionBufferUsageCmd>();
```

Comment:
This should enqueue a transition command in the current command list. You can add a helper function in the Buffer that given a command list enqueues a transition and call it both here and in d3d12::Buffer::TransitionUsageImpl. Or even better make it return the D3D12_RESOURCE_BARRIER so that it doesn't have side effects.